pkg/dpm/manager.go

```diff
@@ -3,99 +3,217 @@ package dpm
-	for deviceClass, deviceIDs := range *devices {
-		dpm.plugins = append(dpm.plugins, lister.NewDevicePlugin(deviceClass, deviceIDs))
+	// Add new plugins
+	for newPluginLastName, _ := range newPluginsSet {
+		wg.Add(1)
+		go func() {
+			if _, ok := currentPluginsMap[newPluginLastName]; !ok {
+				glog.V(3).Infof("Adding a new plugin \"%s\"", newPluginLastName)
+				plugin := newDevicePlugin(dpm.lister.GetResourceNamespace(), newPluginLastName, dpm.lister.NewPlugin(newPluginLastName))
+				startPlugin(newPluginLastName, plugin)
+				pluginMapMutex.Lock()
+				currentPluginsMap[newPluginLastName] = plugin
+				pluginMapMutex.Unlock()
+			}
+			wg.Done()
+		}()
 	}
+	wg.Wait()
+
+	// Remove old plugins
+	for pluginLastName, plugin := range currentPluginsMap {
+		wg.Add(1)
+		go func() {
+			if _, ok := newPluginsSet[pluginLastName]; !ok {
+				stopPlugin(pluginLastName, plugin)
+				pluginMapMutex.Lock()
+				delete(currentPluginsMap, pluginLastName)
+				pluginMapMutex.Unlock()
+			}
+			wg.Done()
+		}()
+	}
+	wg.Wait()
+}
 
-	return dpm
+func (dpm *Manager) startPluginServers(pluginMap map[string]devicePlugin) {
+	var wg sync.WaitGroup
+
+	for pluginLastName, plugin := range pluginMap {
+		wg.Add(1)
+		go func() {
+			startPluginServer(pluginLastName, plugin)
+			wg.Done()
+		}()
+	}
+	wg.Wait()
 }
 
-// Run starts the DevicePluginManager.
-func (dpm *DevicePluginManager) Run() {
-	defer dpm.fsWatcher.Close()
-	dpm.startPlugins()
+func (dpm *Manager) stopPluginServers(pluginMap map[string]devicePlugin) {
+	var wg sync.WaitGroup
 
-	// The manager cannot be restarted, it is expected to be the `core` of an application.
-	<-dpm.stopCh
-	dpm.stopPlugins()
+	for pluginLastName, plugin := range pluginMap {
+		wg.Add(1)
+		go func() {
+			stopPluginServer(pluginLastName, plugin)
+			wg.Done()
+		}()
+	}
+	wg.Wait()
 }
 
-// startPlugins is helper function to start the underlying gRPC device plugins.
-func (dpm *DevicePluginManager) startPlugins() {
-	for _, plugin := range dpm.plugins {
-		go plugin.Start()
+func (dpm *Manager) stopPlugins(pluginMap map[string]devicePlugin) {
+	var wg sync.WaitGroup
+	var pluginMapMutex = &sync.Mutex{}
+
+	for pluginLastName, plugin := range pluginMap {
+		wg.Add(1)
+		go func() {
+			stopPlugin(pluginLastName, plugin)
+			pluginMapMutex.Lock()
+			delete(pluginMap, pluginLastName)
+			pluginMapMutex.Unlock()
+			wg.Done()
+		}()
 	}
+	wg.Wait()
 }
 
-// startPlugins is helper function to stop the underlying gRPC device plugins.
-func (dpm *DevicePluginManager) stopPlugins() {
-	for _, plugin := range dpm.plugins {
-		plugin.Stop()
+func startPlugin(pluginLastName string, plugin devicePlugin) {
+	if devicePluginImpl, ok := plugin.DevicePlugin.(PluginInterfaceStart); ok {
+		err := devicePluginImpl.Start()
+		if err != nil {
+			glog.Errorf("Failed to start plugin \"%s\": %s", pluginLastName, err)
+		}
 	}
+	startPluginServer(pluginLastName, plugin)
```

Comment:
Should the pluginServer start even if the plugin itself failed to start?

Reply:
Done